### ISSUE TITLE
gather inspiration from the best flakes

### ISSUE THREAD

**Issue by dminca** [dminca/nix-config#4]
Seen a lot of professionally written flakes, to name a few
- https://git.sr.ht/~afontaine/home/tree
- https://git.sr.ht/~bwolf/dotfiles/tree
- https://github.com/Mic92/dotfiles
- https://github.com/konradmalik/dotfiles - Konrad even enabled some CI/CD within his configuration with GitHub Actions; might help on #2 
  - https://konradmalik.com/posts/2023/02/sops-nix-simple-secrets-management-for-nix/
- https://github.com/EmergentMind/nix-config
  - https://unmovedcentre.com/technology/2024/03/22/secrets-management.html
- example structure for multihost configuration
  - https://github.com/mwdavisii/nyx/tree/main/home
  - https://github.com/BirdeeHub/birdeeSystems/blob/582fe0c1123395c8cc0aa3a1bf6dfa3ce65dcfbb/flake.nix#L91-L115
  - https://github.com/BirdeeHub/birdeeSystems/tree/main/systems/PCs
  - https://github.com/lucasfabre/nixos-config/tree/main/system
  - https://github.com/mwdavisii/nyx
  - https://github.com/mwdavisii/nyx/tree/main/home
  - 🌟 https://github.com/infinisil/system
- https://github.com/gytis-ivaskevicius/flake-utils-plus
- https://github.com/reckenrode/nixos-configs

In addition, if we're to define `options` within the flake, here's a list of supported `types` by Nix 
- https://github.com/NixOS/nixpkgs/blob/ea24372d132cf854937ef41d0d4091ca0e8f7ee9/lib/types.nix#L224-L1049
- [Official Nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/)

**Comment by dminca** [comment]
Some conversation from the forum of people trying to modularize home-manager configuration 

- [Is this a good way to modularize home manager home.nix for home/work?](https://discourse.nixos.org/t/is-this-a-good-way-to-modularize-home-manager-home-nix-for-home-work/5817)

**Comment by dminca** [comment]
- [Conditional Implementation](https://nixos.wiki/wiki/Extend_NixOS#Conditional_Implementation)

```nixos
 {config, pkgs, lib, ...}:
 
 {
   systemd.services = lib.mkIf (config.networking.hostName == "my-server") {
      ircSession = {
        wantedBy = [ "multi-user.target" ]; 
        after = [ "network.target" ];
        description = "Start the irc client of username.";
        serviceConfig = {
          Type = "forking";
          User = "username";
          ExecStart = ''${pkgs.screen}/bin/screen -dmS irc ${pkgs.irssi}/bin/irssi'';         
          ExecStop = ''${pkgs.screen}/bin/screen -S irc -X quit'';
        };
      };
   };
 
   environment.systemPackages = lib.mkIf (config.networking.hostName == "my-server") [ pkgs.screen ];
 
   # ... usual configuration ...
 }
```

**Comment by dminca** [comment]
- [Module System and Custom Options](https://nixos-and-flakes.thiscute.world/other-usage-of-flakes/module-system)

```nixos
# ./foo.nix
{ config, lib, pkgs, ... }:

with lib;

let
  cfg = config.programs.foo;
in {
  options.programs.foo = {
    enable = mkEnableOption "the foo program";

    package = mkOption {
      type = types.package;
      default = pkgs.hello;
      defaultText = literalExpression "pkgs.hello";
      description = "foo package to use.";
    };

    extraConfig = mkOption {
      default = "";
      example = ''
        foo bar
      '';
      type = types.lines;
      description = ''
        Extra settings for foo.
      '';
    };
  };

  config = mkIf cfg.enable {
    home.packages = [ cfg.package ];
    xdg.configFile."foo/foorc" = mkIf (cfg.extraConfig != "") {
      text = ''
        # Generated by Home Manager.

        ${cfg.extraConfig}
      '';
    };
  };
}
```

**Comment by dminca** [comment]
- [Option definitions are generally straight-forward bindings of values to option names](https://nixos.org/manual/nixos/stable/#sec-option-definitions)

```nixos
config = if config.services.httpd.enable then {
  environment.systemPackages = [ ... ];
  ...
} else {};

```

**Comment by dminca** [comment]
asked [the Discord community](https://discord.com/channels/568306982717751326/1246173068309823509) for proposals and I've got this
- https://git.proot.pl/Wroclaw/nixos-configuration

**Comment by dminca** [comment]
Solved in #7
